bmt/bmt.go

```diff
@@ -0,0 +1,565 @@
+// the BMTree is 'locked' while not in the pool
+type BMTree struct {
+	leaves []*BMTNode
+}
+
+// draws the BMT (badly)
+func (self *BMTree) Draw(hash []byte, d int) string {
+	var left, right []string
+	var anc []*BMTNode
+	for i, n := range self.leaves {
+		left = append(left, fmt.Sprintf("%v", hashstr(n.left)))
+		if i%2 == 0 {
+			anc = append(anc, n.parent)
+		}
+		right = append(right, fmt.Sprintf("%v", hashstr(n.right)))
+	}
+	anc = self.leaves
+	var hashes [][]string
+	for l := 0; len(anc) > 0; l++ {
+		var nodes []*BMTNode
+		hash := []string{""}
+		for i, n := range anc {
+			hash = append(hash, fmt.Sprintf("%v|%v", hashstr(n.left), hashstr(n.right)))
+			if i%2 == 0 && n.parent != nil {
+				nodes = append(nodes, n.parent)
+			}
+		}
+		hash = append(hash, "")
+		hashes = append(hashes, hash)
+		anc = nodes
+	}
+	hashes = append(hashes, []string{"", fmt.Sprintf("%v", hashstr(hash)), ""})
+	total := 60
+	del := "                             "
+	var rows []string
+	for i := len(hashes) - 1; i >= 0; i-- {
+		var textlen int
+		hash := hashes[i]
+		for _, s := range hash {
+			textlen += len(s)
+		}
+		if total < textlen {
+			total = textlen + len(hash)
+		}
+		delsize := (total - textlen) / (len(hash) - 1)
+		if delsize > len(del) {
+			delsize = len(del)
+		}
+		row := fmt.Sprintf("%v: %v", len(hashes)-i-1, strings.Join(hash, del[:delsize]))
+		rows = append(rows, row)
+
+	}
+	rows = append(rows, strings.Join(left, "  "))
+	rows = append(rows, strings.Join(right, "  "))
+	return strings.Join(rows, "\n") + "\n"
+}
+
+// initialises the BMTree by building up the nodes of a BMT
+// segment size is stipulated to be the size of the hash
+// segmentCount needs to be positive integer and does not need to be
+// a power of two and can even be an odd number
+// segmentSize * segmentCount determines the maximum chunk size
+// hashed using the tree
+func NewBMTree(hasher Hasher, segmentSize, segmentCount int) *BMTree {
+	n := NewBMTNode(0, 0, nil)
+	n.root = true
+	prevlevel := []*BMTNode{n}
+	// iterate over levels and creates 2^level nodes
+	level := 1
+	count := 2
+	for d := 1; d <= depth(segmentCount); d++ {
+		nodes := make([]*BMTNode, count)
+		for i, _ := range nodes {
+			var parent *BMTNode
+			parent = prevlevel[i/2]
+			t := NewBMTNode(level, i, parent)
+			// fmt.Printf("created node level %v, index: %v/%v\n", level, i, count)
+			nodes[i] = t
+		}
+		prevlevel = nodes
+		level++
+		count *= 2
+	}
+	// the datanode level is the nodes on the last level where
+	return &BMTree{
+		leaves: prevlevel,
+	}
+}
+
+// methods needed by hash.Hash
+
+// returns the size
+func (self *BMTHasher) Size() int {
+	return self.size
+}
+
+// returns the block size
+func (self *BMTHasher) BlockSize() int {
+	return self.blocksize
+}
+
+// hash.Hash interface Sum method appends the byte slice to the underlying
+// data before it calculates and returns the hash of the chunk
+func (self *BMTHasher) Sum(b []byte) (r []byte) {
+	t := self.getTree()
```

Comment:
this is not needed if Reset takes case of reserving a tree from the pool and assigning to `self.bmt`
same with ReadFrom